dsl/flatten.cc

```diff
@@ -0,0 +1,322 @@
+            level--;
+        }
+
+        methods.methods[methods.stack.back().idx] = std::move(expr);
+        methods.stack.pop_back();
+        return make_unique<ast::EmptyTree>();
+    };
+
+    unique_ptr<ast::Expression> addMethods(core::Context ctx, unique_ptr<ast::Expression> tree) {
+        auto &methods = curMethodSet().methods;
+        if (methods.empty()) {
+            ENFORCE(popCurMethodDefs().empty());
+            return tree;
+        }
+        if (methods.size() == 1 && (ast::cast_tree<ast::EmptyTree>(tree.get()) != nullptr)) {
+            // It was only 1 method to begin with, put it back
+            unique_ptr<ast::Expression> methodDef = std::move(popCurMethodDefs()[0]);
+            return methodDef;
+        }
+
+        auto insSeq = ast::cast_tree<ast::InsSeq>(tree.get());
+        if (insSeq == nullptr) {
+            ast::InsSeq::STATS_store stats;
+            tree = make_unique<ast::InsSeq>(tree->loc, std::move(stats), std::move(tree));
+            return addMethods(ctx, std::move(tree));
+        }
+
+        for (auto &method : popCurMethodDefs()) {
+            ENFORCE(method != nullptr);
+            insSeq->stats.emplace_back(std::move(method));
+        }
+        return tree;
+    }
+
+private:
+    ast::ClassDef::RHS_store addMethods(core::Context ctx, ast::ClassDef::RHS_store rhs) {
+        if (curMethodSet().methods.size() == 1 && rhs.size() == 1 &&
+            (ast::cast_tree<ast::EmptyTree>(rhs[0].get()) != nullptr)) {
+            // It was only 1 method to begin with, put it back
+            rhs.pop_back();
+            rhs.emplace_back(std::move(popCurMethodDefs()[0]));
+            return rhs;
+        }
+        for (auto &method : popCurMethodDefs()) {
+            ENFORCE(method.get() != nullptr);
+            rhs.emplace_back(std::move(method));
+        }
+        return rhs;
+    }
+
+    vector<unique_ptr<ast::Expression>> popCurMethodDefs() {
+        auto ret = std::move(curMethodSet().methods);
+        ENFORCE(curMethodSet().stack.empty());
+        popCurMethodSet();
+        return ret;
+    };
+
+    int computeStaticLevel(ast::MethodDef *methodDef) {
+        auto &methods = curMethodSet();
+        int prevLevel = methods.stack.size() > 0 ? methods.stack.back().staticLevel : 0;
+        return prevLevel + (methodDef->isSelf() ? 1 : 0);
+    }
+
+    struct MethodData {
+        // This is an index into the methods stack
+        int idx;
+        // we need to keep information around about whether we're in a static outer context: for example, if we have
+        //
+        //   def self.foo; def bar; end; end
+        //
+        // then we should flatten it to
+        //
+        //  def self.foo; end
+        //  def self.bar; end
+        //
+        // which means when we get to `bar` we need to know that the outer context `foo` is static. We pass that down
+        // the current stack by means of this `isStatic` variable.
+        int staticLevel;
```

Comment:
This `staticLevel` variable will become a bool (`isStatic`) when we implement the correct logic, which is effectively binary instead of requiring arbitrary nesting.

Reply:
Looking over this, I don't think this comment is very clear, so just to be extra-explicit: the correct way to treat
```ruby
def self.a
  def self.b; end
end
```
is to flatten it to
```ruby
def self.a; end
def self.b; end
```
but our existing logic treated it as though it meant
```ruby
def self.a; end
class << self; def self.b; end; end
```
If I had nested another method inside the definition of `b`, then our current semantics would have treated it as yet another level nested deep. That means that flattening that matches our existing logic needs to includes `class << self` blocks nested as deep as nested static methods. However, the correct semantics don't require arbitrary nesting: a class method defined locally inside a class method is just a class method. Consequently, when we update this to match the correct semantics, all we'll need to store is whether or not the method is static, and now how many levels deep in static methods it was.